### ISSUE TITLE
Add match statement support

### ISSUE THREAD

**Issue by rarajabs** [robotframework/robotframework#5226]
Hello everyone,

I would like to suggest to add match statement in robot framework. 

Match statement is support in python [Match Statement](https://docs.python.org/3/tutorial/controlflow.html#match-statements)

This suggestion is to add another method besides IF / ELSE block.

**Comment by pekkaklarck** [comment]
I've been thinking that MATCH syntax similar to the one in Python would be nice. It would simplify certain validations considerably and would thus be very useful in an automation framework like Robot. I guess the syntax should be something like below:

```
MATCH    ${something}
CASE    Something bad happened.
    Log    That's fine, we'll recover...
CASE    Something really bad happended!!!1!
    Fail    Cannot recover!
CASE
    Log    No match.
END
```
Interestingly you can already nowadays simulate this with TRY/EXCEPT:
```robotframework
*** Test Cases ***
Example
    VAR    ${something}    Something bad happened!
    TRY
        Fail    ${something}
    EXCEPT    Something bad happened!
        Log    That's fine, we'll recover...
    EXCEPT    Something really bad happended!!!1!
        Fail    Cannot recover!
    EXCEPT
        Log    No match.
    END
```
Although TRY/EXCEPT already works, a dedicated MATCH syntax would be fine and make the usage more explicit and avoid the somewhat strange failing TRY block. Someone just needed to design how the MATCH syntax should work in detail first. We can use this issue as a place for discussion, but it might be better to start a separate thread on the #devel channel on our Slack. Even then all design decisions and other important points should be documented here.

**Comment by pekkaklarck** [comment]
A simple way to implement the MATCH syntax would be reusing most of the TRY/EXCEPT syntax for that. We could even do it so that only the TRY block would be replaced with MATCH and everything else would be the same as with TRY/EXCEPT:
```
MATCH    ${something}
EXCEPT    First candidate.
    Log    First candidate matched!
EXCEPT    Second *.    AS    ${value}    type=glob
    Log    Second candidate with value '${value}' matched!
EXCEPT
    Log    No explicit candidate matched.
END
```
Although the above would work, there are some problems with it:
- Should we use plain EXCEPT or ELSE for the case where no explicit EXCEPT match? The above a plain EXCEPT, but perhaps ELSE would read better.
- Regardless would we use plain EXCEPT else ELSE for the default case, we need to decide what what to do with the other. Probably it should be disallowed altogether.
- What to do with FINALLY? Probably should be disallowed.
- Should we use CASE instead EXCEPT as in Python? I think both read well and I'd kind of prefer EXCEPT if we could reuse TRY/EXCEPT syntax otherwise, but because all TRY/EXCEPT blocks probably don't makes sense in a MATCH construct anyway, using EXCEPT could be confusing.

**Comment by pekkaklarck** [comment]
One major design decision is handling non-string values. TRY/EXCEPT is solely string based, but with MATCH we most likely wanted to support lists, dics and possibly also custom objects. We probably also wanted to support capturing certain elements from these constructs like the fist list item or a specific attribute value from an objects. The question is, how should that work in practice?

Something like this could work with lists, but implementing it would require quite a lot of work. We'd also needed to separate implementation for dics, objects, and anything else we wanted to support.
```
MATCH    ${items}
CASE    []
    Log    No items.
CASE    [${item}]
    Log    One item '${item}'.
CASE    ['Hello', ${item}]
    Log    Two items 'Hello' and '${item}'.
CASE    [${item1}, ${item2}]
    Log    Two times '${item1}' and '${item2}'.
CASE    [@{items}]
    Log    Multiple items: ${items}
END
```

**Comment by pekkaklarck** [comment]
As I wrote in my first comment above, I'd like to have this kind of a construct, but as the latter comments have shown, there are various design decisions to be made and especially capturing parts of the match will also be rather complicated to implement. Adding parser support for MATCH/CASE, adding needed execution and result objects, handling this during execution and reporting it properly is also lot of work altogether, but there shouldn't be anything really complicated.

I personally won't have time to look at this more closely in the near future and believe this could be added earliest in RF 8.0 anyway. If someone wants to help with this issue, the first step would be coming up with a syntax proposal. You can also start a discussion about this on the #devel channel on our Slack.

**Comment by mikeleppane** [comment]
Hey! I've now added a basic implementation for the MATCH statement. See the attached zip files. Now that I understand how the robot "interpreter" works, we can start thinking about the syntax and how this should work in the Robot context. As mentioned above, we should probably support lists, dicts, and some other types. In addition, what about default case/catch-all case, AS pattern, use a regex for a capture pattern, etc.. 

Here are two runs for basic valid and invalid cases using the MATCH statement.

[valid_results.zip](https://github.com/user-attachments/files/17542332/valid_results.zip)

[invalid_results.zip](https://github.com/user-attachments/files/17542334/invalid_results.zip)

**Comment by mikeleppane** [comment]
Yeah, I don't see the benefit of creating a draft PR at this point. But I will keep experimenting, and the next thing I will do is add support for the sequence pattern. While experimenting, the correct syntax will become more evident and, in addition, how complex is it actually to support non-string values? 

Syntax proposal:

```python
MATCH    <subject>
CASE    <pattern>    type=GLOB|REGEXP    AS    ${capture}    // Inline IF would be great here => IF <condition> 
    KEYWORD   <args>
CASE    <pattern>
    KEYWORD   <args>
...
DEFAULT
    KEYWORD   <args>
```

**Comment by mikeleppane** [comment]
@pekkaklarck, here's the current progress regarding handling sequence patterns (see the attachment file).
You are right; `AS  ${capture}` is inadequate. Furthermore, do we even need the whole `AS` pattern? 
The next thing would be to add support for mapping patterns. 

> You probably often want to capture a part from a string
Do you mean that we should have something as follows:
```
CASE    Error: ${rest}
Log    ${rest}
```    

[sequence_type_results.zip](https://github.com/user-attachments/files/17606632/sequence_type_results.zip)

**Comment by mikeleppane** [comment]
> or a specific attribute value from an objects.
How should we support matching objects (class pattern) in a Robot context? This is somewhat of a question mark for me. 
I have added support for something like this if it makes sense:
```python
MATCH    ${value}
CASE    int()
    # Do something
CASE    int('${value}') 
    Log   ${value}
CASE    list()
    # Do something
CASE    bool()
    # Do something
CASE    str()
    # Do something
END
```
Also capturing a value is possible:
```python
MATCH    ${value}
CASE    int('${value}') 
    Log    ${value}
CASE    str('${text}')
    Log    ${'${text}'}
END
```

**Comment by mikeleppane** [comment]
I think the current implementation is starting to take shape, so it would be good to go over the syntax we want. Below is a summary of what is now supported and the syntax used. It's good to have a consensus about the syntax to prevent any waste. After that, it's easier for me to start wrapping things up (code clean-up, overall fine-tuning, improving testing) and make the first draft (PR?) version available.

Syntax proposal:
```python
MATCH    <subject>
CASE    <pattern>    IF <condition>    AS <capture>
    KEYWORD   ${capture} 
CASE    <pattern>
    KEYWORD   <args>
...
DEFAULT    AS    <capture>
    KEYWORD   <args>
END
```

#### Literal/value/capture pattern
```python
MATCH    <value>
CASE    1    
CASE    >
CASE    ''    # empty
END
```

```python
# capture
MATCH    Robot Framework
CASE    ${string} Framework
    Log    ${string} Framework
CASE    Robot ${string}
    Log    Robot ${string}
CASE     ${string}
    Log    ${string}
END
```
#### Sequence pattern
```python
@{seq} =    Create List    ${1}    ${2}    ${3}
MATCH    ${seq}
CASE    []
    # empty
CASE    [1]
    ...
CASE    [1,2,3]
    # match
CASE   ['${item}', 2, 3]
    # take first item
CASE   ['${item}', '${item2}', 3]
    # take the first and second items
CASE    ['@{items}']
    # all items
END
```
Also capturing nested lists and sub-groups is possible.

#### Mapping pattern
```python
&{dict} =    Create Dictionary    key=${1}    key2=${2}    key3=${True}
MATCH    ${dict}
CASE    {}
    # empty
CASE     {'key': 1, 'key2': 2, 'key3': True}
    # match
CASE    {'key': '${val}', 'key2': '${val2}', 'key3': True}
    # take the 'key' value and 'key2' value
END
```

#### Class pattern
```python
VAR    ${value} =    ${1.5}
MATCH    ${value}
CASE    str()
    ...
CASE    float()
    ...
CASE    bool()
    ...
CASE    list()
    ...
CASE    dict()
    ...
END
```
I'm still somewhat a question mark whether we need this  `AS <capture>` pattern, but perhaps it would be good for someone?

**Comment by pekkaklarck** [comment]
Thanks for the proposal. Unfortunately there are various cases where the syntax is too different to what Robot otherwise supports. The biggest problem is automatic conversion of values and I explain it below.

The examples contained cases like this:
```
CASE    1      # This apparently matches integer 1.
CASE    '1'    # And this matches string 1.
CASE    ''     # And this matches an empty string.
```

All this is different to how Robot normally handles data. Normally `1` is a string one, `'1'` is a three character string containing `'`, `1` and `'`, and `''` is a two character string containing two times `'`. If you need anything else than literal string values, you need to use variables or the keyword accepting the argument needs to have appropriate type hints for Robot to do automatic argument conversion. I don't think we can use different approach with CASE statements. The biggest reasons are:
- Using two set of rules is confusing. Related to that, the existing rules are complicated enough.
- If `'1'` would match string `1`, what should be used for matching literal `'1'`? I guess there could be some escaping mechanism, but it would make this even more complicated.

An alternative to require all non-string values to be specified as variables would follow existing rules. The above examples would then be written like this:
```
CASE    ${1}         # Integer 1.
CASE    1            # String 1.
CASE    ${EMPTY}     # An empty string.
```

Unfortunately that wouldn't work too well with more complex objects. It would be possible to use inline-Python syntax, but this isn't pretty:
```
CASE    ${{[1, 2, 3]}}
CASE    ${{{'a': 1, 'b': 2, 'c': 3}}}
```

Another possibility would be using automatic conversion based on the type of the matched value. For example, if `${rc}` would be an integer, values with each CASE would be converted to an integer automatically:
```
MATCH    ${rc}
CASE    0
CASE    1
CASE    42
CASE    255
```

To make it more flexible, we could allow specifying the type. It could be specified with each CASE, but I believe specifying it once with MATCH would be better. For example, this would try conversion to a `list[int]` first and if that fails it would try `dict[str, int]`:
```
MATCH    ${json}    type=list[int] | dict[str, int]
CASE    [1, 2, 3]
CASE    {'a': 1, 'b': 2}
```

If an explicit type would be specified, it probably would make sense to validate that the matched value matches the type. A conversion failure with any CASE should also be an error. You could add `| Any` at the end to allow any value.

**Comment by pekkaklarck** [comment]
The proposed syntax for class patterns is problematic as well:
```
CASE    str()
CASE    int()
```
This would make it impossible to match a literal value `str()` and all future class pattern additions could break existing literal patterns. In addition to that, `int()` isn't a syntax we use anywhere else in Robot's data. Finally, this wouldn't support more complex types like a list of integers.

I believe a better syntax would be this:
```
CASE    type=str
CASE    type=int
CASE    type=list[int]
```
The explicit `type=` prefix makes it clear that matching is based on type. The syntax is same as type hints in Python that libraries already support and that we are planning to introduce to variables in Robot data as well.

To allow matching a literal value `type=str`, we need to reserve another prefix like `value=`. Then it would be possible to use `value=type=str` to match a literal value `type=str`.

**Comment by pekkaklarck** [comment]
My proposals above ought to handle base types, including containers, pretty well, but we needed to have some syntax for matching custom objects as well. One possibility would be something like this:
```
CASE    object=User(name='Robot', password='secret')
```
This would match if the value has a type `User` and it contains attributes `name` and `password` with values `Robot` and `secret`, respectively.

This probably isn't a mandatory feature in the first version, so we can decide to forget about it for now. Something certainly can be implemented later with a custom prefix.

**Comment by pekkaklarck** [comment]
One hard part is agreeing on the syntax for capturing values. The `AS` syntax that EXCEPT supports probably isn't useful with CASE, because you often want to capture parts of the value. Using variables would probably be the best solution:
```
CASE    Hello, ${name}!          # A string with part of it being captured to a variable.
CASE    [${first}, ${second}]    # List with two items, both captured to variables.
```
A problem with this is that it looks like the value used with CASE would be constructed from variables. I guess users would get used to it, and we also use a similar syntax with keyword accepting embedded arguments. We nevertheless needed to decide should we somehow allow constructing these values from variables. If it's considered useful, we needed to have some syntax for it.

A technical problem is that implementing this properly is likely pretty complicated. Handling strings ought to be easy as we probably could reuse lot of the code related to embedded arguments, but handling list and dictionaries is much harder. I don't think string matching would work well, and instead we should tokenize them ourselves and go through items one-by-one. This can be tricky especially with nested containers.

**Comment by pekkaklarck** [comment]
Another challenge is handling wild cards. With strings we needed to have some way to match only parts of it. One alternative would be using a dedicated prefix to specify that the value is a regexp:
```
CASE    regexp=Beginning.*End
CASE    regexp=Hello, ${name}[!.]?
```
We could then use also `glob=` if we wanted to support simpler glob patterns. Alternatively we could use a dedicated syntax like `/pattern/`, but then we needed to have a way to escape it. We have the same needed with embedded arguments, so whatever solution we'd choose, should work in both usages.

Regexp or glob based solutions wouldn't work with containers. I believe with them we should allow something like `*`, `_` or `...` to denote optional items:
```
CASE    [1, 2, 3, ..., 10]       # List starting with 1, 2, 3 and ending with 10.
CASE    {'a': 1, 'b': 2, ...}    # Dictionary with at least items 'a': 1 and 'b': 2.
```
Handling containers like this will require tokenizing them and makes this pretty complicated. We'd anyway need to do that to handle capturing items from containers, though.

**Comment by pekkaklarck** [comment]
I started a discussion about this on the #devel channel on our Slack. If you @mikeleppane aren't there yet, please join!

**Comment by kstine** [comment]
> Some random comments:
> 
> - I don't think a separate DEFAULT branch is a good idea. Better make it possible to have CASE that matches anything. Possibly the value could be left totally empty like we support with EXCEPT.
> 
> - I guess it would be good to support multiple values with one CASE, but it doesn't need to be supported in the first version. The syntax should be just listing values so that they are separated normally with spaces, no OR is needed. We use that approach also with EXCEPT.
> 
> 

We could adopt the python default syntax of `CASE _`

**Comment by mikeleppane** [comment]
Great. I mostly agree with what you explain above. Here are some comments:

> Another possibility would be using automatic conversion based on the type of the matched value. For example, if ${rc} would be an integer, values with each CASE would be converted to an integer automatically:

I approached this like this: using automatic conversion without needing a user to specify any `type=<type>`.  And what if the list is more complicated than just a list of ints? 

> To make it more flexible, we could allow specifying the type. It could be specified with each CASE, but I believe specifying it once with MATCH would be better. For example, this would try conversion to a list[int] first and if that fails it would try dict[str, int]:

Yes, I guess we would this `type` prefix, and I like this class pattern syntax better:
```python
CASE    type=str  # instead str()
```

> This probably isn't a mandatory feature in the first version, so we can decide to forget about it for now. Something certainly can be implemented later with a custom prefix.

Yeah, let's leave this open at this point. But the `object` prefix is one possibility.

> The AS syntax that EXCEPT supports probably isn't useful with CASE, because you often want to capture parts of the value. 
I agree. I think we probably don't need `AS` syntax for the CASE clause

> A technical problem is that implementing this properly is likely pretty complicated. Handling strings ought to be easy as we probably could reuse lot of the code related to embedded arguments, but handling list and dictionaries is much harder. I don't think string matching would work well, and instead we should tokenize them ourselves and go through items one-by-one. This can be tricky especially with nested containers.

Well, the way it's now implemented (e.g., sequence pattern) is that a `<pattern>` is converted to a Python object. This somewhat simplifies the data handling even though the algorithm is far from trivial. But the downside is that this requires moving away from the native Robot syntax. E.g.
```python
CASE    [${first}, ${second}]  # vs
CASE    ['${first}', '${second}'] 
```
But again, this is probably not the correct path to go. I'm not sure how much more complicated the current algorithm becomes if we need to parse/tokenize pattern input. And also we would need to find alternative solution wildcard pattern.

> I believe with them we should allow something like *, _ or ... to denote optional items:

I like this. I need to try out how would it work...

> Another challenge is handling wild cards. With strings we needed to have some way to match only parts of it. One alternative would be using a dedicated prefix to specify that the value is a regexp:

So, we would need these `regexp/glob=` prefixes?



> I don't think a separate DEFAULT branch is a good idea. Better make it possible to have CASE that matches anything. Possibly the value could be left totally empty like we support with EXCEPT.

Why don't you think DEFAULT is a good idea? What's the alternative when we don't have a good keyword for matching everything like in Python and Rust lands. I mean `CASE` w/o anything reads badly and is not very intuitive.   

> I guess it would be good to support multiple values with one CASE, but it doesn't need to be supported in the first version. The syntax should just list values so that they are separated normally with spaces; no OR is needed. We use that approach also with EXCEPT.

I guess it's possible to list pattern-separated space, but I'm suspicious about how well it reads. What would it look like:
```python
CASE    1 2 3 # vs
CASE    1 OR 2 OR 3
```

**Comment by pekkaklarck** [comment]
Quick answers:
1. I understand `CASE    ['${first}', '${second}']` would be easier to support than `CASE    [${first}, ${second}]`, but I believe the latter is anyway the right syntax to use. It will most likely require a separate tokenizer, but I rather not have this support than have syntax that's not optional. That said, we can consider first implementing support only for stings and possibly other base types, and look at containers later. Even in that case we needed to thing about the syntax so that we wouldn't need to change existing syntax when the new container syntax would be added. 
2. I don't think it makes sense to create a separate token, parsing node, execution object and result object for the default case. For example, Python uses `case _` instead and we already use an empty `EXCEPT` to catch any exception. I think an empty `CASE` would be fine especially because it would be consistent with `EXCEPT`, but we could also consider something like `CASE    *` or `CASE    DEFAULT` that doesn't require a separate token, node, etc.
3. Multiple values with `CASE` would be written like `CASE    value1    value2`. That's consistent with `EXCEPT    Error 1    Error 2` that we already support. Using `OR` with one but not with the other is not an option and realistically changing the `EXCEPT` syntax to support `OR` isn't really an option either.

**Comment by pekkaklarck** [comment]
By the way, if you @mikeleppane want to work with a simpler enhancement requiring parser etc. changes first, the proposed BLOCK syntax (#5257) would have basically no open design decisions. I think it would be a nice addition, especially because it would make it easy to create nested structures programmatically, but let's see what others here or on #devel think about it first.

**Comment by mikeleppane** [comment]
Yeah, I've already started looking into this `BLOCK` scope thing and a very basic version is running on my computer.
